src/core/thread/link_quality.cpp

```diff
@@ -36,170 +36,149 @@
 #include "link_quality.hpp"
 
 #include <stdio.h>
-#include "utils/wrap_string.h"
 
 #include "common/code_utils.hpp"
+#include "utils/wrap_string.h"
 
 namespace ot {
 
-enum
+// This array gives the decimal point digits representing 0/8, 1/8, ..., 7/8 (does not include the '.').
+static const char *const kDigitsString[8] =
 {
-    kDefaultNoiseFloor = -100, // Default noise floor used if no average value is available.
+    // 0/8,  1/8,   2/8,   3/8,   4/8,   5/8,   6/8,   7/8
+    "0",     "125", "25",  "375", "5",   "625", "75",  "875"
 };
 
-// This array gives the decimal point digits representing 0/8, 1/8, ..., 7/8 (it does not include the '.').
-static const char *const kLinkQualityDecimalDigitsString[8] =
-{
-    // 0/8, 1/8,   2/8,   3/8,   4/8,   5/8,   6/8,   7/8
-    "000", "125", "250", "375", "500", "625", "750", "875"
-};
-
-const char LinkQualityInfo::kUnknownRssString[] = "Unknown RSS";
-
-//-------------------------------------------------------------------------------
-
-LinkQualityInfo::LinkQualityInfo(void)
+void RssAverager::Reset(void)
 {
-    Clear();
+    mAverage = 0;
+    mCount = 0;
 }
 
-void LinkQualityInfo::Clear(void)
-{
-    mRssAverage  = 0;
-    mCount       = 0;
-    mLinkQuality = 0;
-    mLastRss     = 0;
-}
-
-void LinkQualityInfo::AddRss(int8_t aNoiseFloor, int8_t aRss)
+otError RssAverager::Add(int8_t aRss)
 {
-    uint16_t    newValue;
-    uint16_t    oldAverage;
-
-    VerifyOrExit(aRss != OT_RADIO_RSSI_INVALID);
-
-    mLastRss = aRss;
+    otError error = OT_ERROR_NONE;
+    uint16_t newAverage;
+    uint16_t oldAverage;
 
-    // Restrict/Cap the RSS value to the closed range [0, -128] so the value can fit in 8 bits.
+    VerifyOrExit(aRss != OT_RADIO_RSSI_INVALID, error = OT_ERROR_INVALID_ARGS);
 
+    // Restrict the RSS value to the closed range [0, -128] so the RSS times precision multiple can fit in 11 bits.
     if (aRss > 0)
     {
         aRss = 0;
     }
 
     // Multiply the the RSS value by a precision multiple (currently -8).
 
-    newValue = static_cast<uint16_t>(-aRss);
-    newValue <<= kRssAveragePrecisionMultipleBitShift;
+    newAverage = static_cast<uint16_t>(-aRss);
+    newAverage <<= kPrecisionBitShift;
 
-    oldAverage = mRssAverage;
+    oldAverage = mAverage;
 
-    if (mCount >= kRssCountForWeightCoefficientOneEighth)
+    if (mCount >= kCountForOneEighth)
     {
         // New average = old average * 7/8 + new value * 1/8
-        mRssAverage = static_cast<uint16_t>(((oldAverage << 3) - oldAverage + newValue) >> 3);
+        mAverage = static_cast<uint16_t>(((oldAverage << 3) - oldAverage + newAverage) >> 3);
```

Comment:
My initial goal was  to avoid using `/` and do all math with `>>` and `<<` to keep it fast/simple. That's why the weights are `1/2, 1/4, 1/8` (basically `1/2^j` that best approximates `1/k`).

I am fine with changing this to `1/k` and use the `/` instead (the radio chips should be fast enough for division calculation).